File: 2015/day4/4.py
import hashlib

from common.utils import problem_harness, timeit, read_input


def mine(key: str = None, difficulty: int = 1) -> int:
  i = 0
  while i < 2**32:
    h = hashlib.md5((key + str(i)).encode()).hexdigest()
    if h[:difficulty] == "0" * difficulty:
      return i
    i += 1
  raise ValueError("No solution found")


@timeit
def part1(filename: str) -> int:
  return mine(read_input(filename)[0], 5)

@timeit
def part2(filename: str) -> int:
  return mine(read_input(filename)[0], 6)


def main():
  problem_harness(part1, part2)

if __name__ == '__main__':
  main()

File: 2015/day4/4.go
package main


import (
	"crypto/md5"
	"fmt"
	"strings"
	"aoc/utils"
)


func mine(key string, difficulty int) int {
	for i := 0; ; i++ {
		hash := hash(fmt.Sprintf("%s%d", key, i))[:difficulty]
		if hash == strings.Repeat("0", difficulty) {
			return i
		}
	}
}


func hash(s string) string {
	return fmt.Sprintf("%x", md5.Sum([]byte(s)))
}


func part1(data string) int {
	key := strings.TrimRight(data, "\n") 
	return mine(key, 5)
}


func part2(data string) int {
	key := strings.TrimRight(data, "\n") 
	return mine(key, 6)
}


func main() {
	utils.ProblemSetUp(part1, part2)
}